src/drivers/intel/cavs/mn.c

```diff
@@ -142,73 +171,190 @@ static bool find_mn(uint32_t freq, uint32_t bclk,
+					       uint32_t *m, uint32_t *n)
+{
+	int clk_index = find_bclk_source(bclk, scr_div, m, n);
+
+	if (clk_index < 0) {
+		trace_mn_error("error: BCLK %d, no valid source", bclk);
+		return -EINVAL;
 	}
 
-	if (clk_index >= 0) {
-		mdivc |= MNDSS(ssp_freq_sources[clk_index]);
-		*out_scr_div = ssp_freq[clk_index].freq / bclk_rate;
-
-		/* select M/N output for bclk in case of Audio Cardinal
-		 * or PLL Fixed clock.
-		 */
-		if (ssp_freq_sources[clk_index] != SSP_CLOCK_XTAL_OSCILLATOR)
-			*out_need_ecs = true;
-	} else {
-		/* check if we can get target BCLK with M/N */
-		for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++) {
-			if (find_mn(ssp_freq[i].freq, bclk_rate,
-				    out_scr_div, &i2s_m, &i2s_n)) {
-				clk_index = i;
-				break;
-			}
-		}
+	bclk_source_mn_clock = clk_index;
+
+	mn_reg_write(0x0, (mn_reg_read(0x0) |
+			   MNDSS(ssp_freq_sources[clk_index])));
+	return 0;
+}
+
+/**
+ * \brief Finds valid M/(N * SCR) values for source clock that is already locked
+ *	  because other ports use it.
+ * \param[in] bclk Bit clock frequency.
+ * \param[out] scr_div SCR divisor.
+ * \param[out] m M value of M/N divider.
+ * \param[out] n N value of M/N divider.
+ * \return 0 on success, error code otherwise.
+ */
+static inline int setup_current_bclk_mn_source(uint32_t bclk, uint32_t *scr_div,
+					       uint32_t *m, uint32_t *n)
+{
+	/* source for M/N is already set, no need to do it */
+	if (find_mn(ssp_freq[bclk_source_mn_clock].freq, bclk, scr_div, m, n))
+		return 0;
+
+	trace_mn_error("error: BCLK %d, no valid configuration for already selected source = %d",
+		       bclk, bclk_source_mn_clock);
+	return -EINVAL;
+}
 
-		if (clk_index < 0) {
-			trace_mn_error("error: BCLK %d", bclk_rate);
-			ret = -EINVAL;
-			goto out;
+#if CAVS_VERSION >= CAVS_VERSION_2_0
+static inline bool check_bclk_xtal_source(uint32_t bclk, bool mn_in_use,
+					  uint32_t *scr_div)
+{
+	/* since cAVS 2.0 bypassing XTAL (ECS=0) is not supported */
+	return false;
+}
+#else
+/**
+ * \brief Checks if XTAL source for BCLK should be used.
+ *	  Before cAVS 2.0 BCLK could use XTAL directly (without M/N).
+ *	  BCLK that use M/N = 1/1 or bypass XTAL is preferred.
+ * \param[in] bclk Bit clock frequency.
+ * \param[in] mn_in_use True if M/N source is already locked by another port.
+ * \param[out] scr_div SCR divisor.
+ * \return true if XTAL source should be used, false otherwise.
+ */
+static inline bool check_bclk_xtal_source(uint32_t bclk, bool mn_in_use,
+					  uint32_t *scr_div)
+{
+	int i;
+
+	for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++) {
+		if (ssp_freq[i].freq % bclk != 0)
+			continue;
+
+		if (ssp_freq_sources[i] == SSP_CLOCK_XTAL_OSCILLATOR) {
+			/* XTAL turned out to be lowest source that can work
+			 * just with SCR, so use it
+			 */
+			*scr_div = ssp_freq[i].freq / bclk;
+			return true;
 		}
 
-		trace_mn("M = %d, N = %d", i2s_m, i2s_n);
+		/* check if can be handled in M/N setup */
```

Comment:
@jajanusz can you please explain what is checked for here? check if "what" can be handled in M/N setup?

Reply:
What is checked is shown in code - if mn is already busy & if it's busy then maybe we can use it cos it's set to desired clock already, if we can, then just quit this function, because code for setting up M/N source (instead of XTAL) will handle it. Comment just says why, what is done I guess is visible in `if` statement. This whole function is just for special case for CAVS<2 where we can bypass XTAL instead of using M/N.

Reply:
Added more descriptive comment